simulator/main.rs

```diff
@@ -67,6 +123,433 @@ enum Value {
+        }
+    }
+}
+
+struct Delete {
+    table: String,
+    predicate: Predicate,
+}
+
+impl ArbitraryFrom<Table> for Delete {
+    fn arbitrary_from<R: Rng>(rng: &mut R, table: &Table) -> Self {
+        Delete {
+            table: table.name.clone(),
+            predicate: Predicate::arbitrary_from(rng, table),
+        }
+    }
+}
+
+impl ArbitraryFrom<Table> for Query {
+    fn arbitrary_from<R: Rng>(rng: &mut R, table: &Table) -> Self {
+        match rng.gen_range(0..=200) {
+            0 => Query::Create(Create::arbitrary(rng)),
+            1..=100 => Query::Select(Select::arbitrary_from(rng, &vec![table])),
+            101..=200 => Query::Insert(Insert::arbitrary_from(rng, table)),
+            // todo: This branch is currently never taken, as DELETE is not yet implemented.
+            //       Change this when DELETE is implemented.
+            201..=300 => Query::Delete(Delete::arbitrary_from(rng, table)),
+            _ => unreachable!(),
+        }
+    }
+}
+
+struct CompoundPredicate(Predicate);
+struct SimplePredicate(Predicate);
+
+impl ArbitraryFrom<(&Table, bool)> for SimplePredicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, (table, predicate_value): &(&Table, bool)) -> Self {
+        // Pick a random column
+        let column_index = rng.gen_range(0..table.columns.len());
+        let column = &table.columns[column_index];
+        let column_values = table.rows.iter().map(|r| &r[column_index]).collect::<Vec<_>>();
+        // Pick an operator
+        let operator = match rng.gen_range(0..3) {
+            0 => {
+                if *predicate_value {
+                    Predicate::Eq(
+                        column.name.clone(),
+                        Value::arbitrary_from(rng, &column_values),
+                    )
+                } else {
+                    Predicate::Eq(
+                        column.name.clone(),
+                        Value::arbitrary_from(rng, &column.column_type),
+                    )
+                }
+            }
+            1 => Predicate::Gt(
+                column.name.clone(),
+                match predicate_value {
+                    true => GTValue::arbitrary_from(rng, &column_values).0,
+                    false => LTValue::arbitrary_from(rng, &column_values).0,
+                },
+            ),
+            2 => Predicate::Lt(
+                column.name.clone(),
+                match predicate_value {
+                    true => LTValue::arbitrary_from(rng, &column_values).0,
+                    false => GTValue::arbitrary_from(rng, &column_values).0,
+                },
+            ),
+            _ => unreachable!(),
+        };
+
+        SimplePredicate(operator)
+    }
+}
+
+impl ArbitraryFrom<(&Table, bool)> for CompoundPredicate {
+    fn arbitrary_from<R: Rng>(rng: &mut R, (table, predicate_value): &(&Table, bool)) -> Self {
+        // Decide if you want to create an AND or an OR
+        CompoundPredicate(if rng.gen_bool(0.7) {
```

Comment:
These random constants would nicer to have extracted somewhere else to understand why chose a specific percentage. These are things where a workload config is nice to be setup from the start, because we can change all of these constants to something else to create a completely different scenario.

Reply:
Yeah agreed. This specific one is rather arbitrary, I can just turn it into an equal probability version.